feat(contrib): add twmb/franz-go integration

igoragoli opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Commit and PR titles should be prefixed with the general area of the pull request's change.

-->
### What does this PR do?

- Adds a [twmb/franz-go](https://github.com/twmb/franz-go?tab=readme-ov-file) integration.

<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succinct form, consider
  opening multiple pull requests instead of a single one.
-->

### Motivation

https://github.com/DataDog/dd-trace-go/issues/2735

<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
* If this resolves a GitHub issue, include "Fixes #XXXX" to link the issue and auto-close it on merge.
-->

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).
-->

- [x] Changed code has unit tests for its functionality at or near 100% coverage.
- [ ] [System-Tests](https://github.com/DataDog/system-tests/) covering this feature have been added and enabled with the va.b.c-dev version tag.
- [ ] There is a benchmark for any new code, or changes to existing code.
- [ ] If this interacts with the agent in a new way, a system test has been added.
- [ ] New code is free of linting errors. You can check this by running `./scripts/lint.sh` locally.
- [ ] Add an appropriate team label so this PR gets put in the right place for the release notes.
- [ ] Non-trivial go.mod changes, e.g. adding new modules, are reviewed by @DataDog/dd-trace-go-guild.

Unsure? Have a question? Request a review!

igoragoli committed:
init carrier.go from local example

igoragoli committed:
copy types.go from segmentio/kafka-go

igoragoli committed:
rename Message to Record

igoragoli committed:
add a TODO for renaming the carrier

igoragoli committed:
don't use kgo, use the structs/interfaces on types.go

igoragoli committed:
simplify comments

igoragoli committed:
fix Set()

igoragoli committed:
copy tracer.go from segmentio/kafka-go

igoragoli committed:
add PackageTwmbFranzGo to packages.go

igoragoli committed:
point to the PackageTwmbFranzGo instrumentation

igoragoli committed:
make comments on tracer.go standardized

igoragoli committed:
copy tracing.go from segmentio/kafka-go

igoragoli committed:
change header and componentName

igoragoli committed:
update StartConsumeSpan

igoragoli committed:
update StartProduceSpan

igoragoli committed:
exporting KafkaHeadersCarrier

igoragoli committed:
tracer.go is now options.go

igoragoli committed:
forgot the constructor

igoragoli committed:
remove question

igoragoli committed:
copy dsm.go from segmentio/kafka-go and edit it

igoragoli committed:
add tracing_test.go

igoragoli committed:
add tracing.go based on the one on segmentio/kafka-go

igoragoli committed:
small fixes on tracing.go

igoragoli committed:
fix internal/tracing path on tracing.go

igoragoli committed:
add franz-go go.mod

igoragoli committed:
init kgo.go with hooks for setting bootstrap servers

igoragoli committed:
add (Get|Set)KafkaConfig on tracing.go

igoragoli committed:
add TODO to make sure I edit writer later

igoragoli committed:
fix interface

igoragoli committed:
update OnProduceRecordBuffered

igoragoli committed:
remove writer from StartProduceSpan interface

igoragoli committed:
update OnProduceRecordUnbuffered

igoragoli committed:
AddBootstrapServer

igoragoli committed:
use AddBootstrapServer on kgo.go

igoragoli committed:
We don't need OnFetchRecordBuffered

igoragoli committed:
update OnFetchRecordUnbuffered

igoragoli committed:
unnecessary imports

igoragoli committed:
set consumer group id

igoragoli committed:
set consumer group id

igoragoli committed:
comment

igoragoli committed:
fix

igoragoli committed:
logs that actually reflect what's being done

igoragoli committed:
dsm

igoragoli committed:
wip

igoragoli committed:
wip

igoragoli committed:
fix go.work and go.mod to enable running tests

igoragoli committed:
add testClient()

igoragoli committed:
wip

igoragoli committed:
wip

igoragoli committed:
a pointer!

igoragoli committed:
test

igoragoli committed:
test

igoragoli committed:
remove dead code

igoragoli committed:
context

igoragoli committed:
remove Tracer.SetConsumerGroupId()

igoragoli committed:
span links

igoragoli committed:
note

igoragoli committed:
refactor: simplify topic mgmt in kgo_test.go

igoragoli committed:
fix: finish spans on close

igoragoli committed:
fix: call consumerCl.Close() on TestProduceConsumeFunctional

igoragoli committed:
refactor: simplify ensureTopicReady()

igoragoli committed:
add franz-go to option.go

igoragoli committed:
test: add more tracing tests

igoragoli committed:
test: add DSM tests

igoragoli committed:
test: explain why we create the expected pathway

igoragoli committed:
test: explain why we create the expected pathway

igoragoli committed:
feat: lazily fetch consumer group ID when polling fetches

igoragoli committed:
docs: update comment to explain why we set consumer group ID lazily

igoragoli committed:
feat: init orchestrion.yml

Snapshot Claude learning mode TODO(human)

igoragoli committed:
tweak: remove bootstrapservers config

igoragoli committed:
test: do not check for bootstrap servers

igoragoli committed:
docs: improve comment formatting on dsm.go

igoragoli committed:
fix: remove unused Writer

genesor commented on contrib/twmb/franz-go/kgo.go:
```diff
@@ -0,0 +1,132 @@
+	c.tracer.FinishProduceSpan(span, int(r.Partition), r.Offset, err)
+}
+
+func (c *Client) OnFetchRecordUnbuffered(r *kgo.Record, polled bool) {
```
Add a comment here explaining that this is a hook, when it is called and what is does.

genesor reviewed (commented):
Great work !

igoragoli committed:
refactor: types.go -> record.go for readability

igoragoli committed:
refactor: move KafkaConfig to tracing.go + docs

wip

igoragoli committed:
refactor: formatting

igoragoli commented on contrib/twmb/franz-go/internal/tracing/record.go:
```diff
@@ -0,0 +1,44 @@
+}
+
+// KafkaConfig holds information from the kafka config for span tags.
+type KafkaConfig struct {
```
Addressed in [3113491](https://github.com/DataDog/dd-trace-go/pull/4250/commits/31134918dd9557079a657992e08eca5327fdf408).

igoragoli committed:
docs: add comments explaining Header/Record interfaces

igoragoli committed:
feat: add GetContext to Record interface

igoragoli committed:
refactor: use record context in DSM checkpoints

igoragoli committed:
refactor: update context before DSM checkpoint

igoragoli committed:
refactor: consolidate getMsgSize functions

igoragoli committed:
refactor: address code review feedback

igoragoli committed:
docs: remove unnecessary 'compile time type assertion' comment

igoragoli committed:
docs: fix copy-pasted Apache license on dsm.go

igoragoli committed:
docs: fix OptionFn docs

igoragoli committed:
docs: remove note i made for learning

igoragoli committed:
feat: add span links to StartProduceSpan

igoragoli committed:
docs: remove unecessary TODO from franz-go instrumentation

igoragoli committed:
docs: improve comment about Kafka header

igoragoli committed:
feat: only continue the propagation of a span if the spanctx is not nil

igoragoli committed:
test: remove unecessary tests

igoragoli committed:
docs: document NewClient

igoragoli committed:
docs: add TODO

igoragoli committed:
feat: remove initial setup of the groupId, since it will be fetched lazily

igoragoli committed:
docs: finishAndClearActiveSpans

igoragoli committed:
docs: go style

igoragoli committed:
docs: improve poll(records|fetches) comments

igoragoli committed:
docs: add kgo hooks comments

igoragoli commented on contrib/twmb/franz-go/kgo.go:
```diff
@@ -0,0 +1,132 @@
+	return c.Client.PollRecords(ctx, maxPollRecords)
+}
+
+func (c *Client) OnProduceRecordBuffered(r *kgo.Record) {
```
Addressed in 1b9107a5.

igoragoli committed:
test: uncomment integration test env var lookup

igoragoli committed:
feat: remove orchestrion.yml

igoragoli committed:
feat: rename tracing.wRecord to tracing.record

igoragoli committed:
docs: add a nice comment explaining tracing.records adapter

igoragoli committed:
feat: rename tracing.wHeader to tracing.header

igoragoli committed:
docs: add a nice comment explaining tracing.header adapter

igoragoli committed:
feat: add NewClientWithTracingOptions()

igoragoli committed:
fix: syntax errors on tests

igoragoli committed:
docs: add periods to doc comments

igoragoli committed:
fix: nil active spans init instead of allocation

igoragoli committed:
tweak: remove unnecessary tag check on TestProduceFunctional

igoragoli committed:
fix: add missing copyright

igoragoli committed:
Merge branch 'main' into augusto/add-twmb-franz-go-integration

igoragoli committed:
refactor: format

igoragoli committed:
fix: modules

igoragoli committed:
fix: blank import to pass TestIntegrationEnabled

igoragoli committed:
fix: option_test.go

igoragoli committed:
refactor: format

igoragoli committed:
fix: contribs_generated

igoragoli committed:
Merge branch 'main' into augusto/add-twmb-franz-go-integration

igoragoli committed:
fix: remove review

rarguelloF reviewed (changes requested):
Thanks so much for the contribution!

I left a request changes since this library provides a tracing hooks option and we don't need to return a custom `Client` type and the internal package. We did this for other integrations (e.g. confluent kafka-go), because they don't provide native options for tracing. Left more details in the comments.

igoragoli committed:
fix: remove unnecessary build tags

igoragoli committed:
feat: replace Client.Close() wrapper with OnClientClosed() hook

igoragoli committed:
Merge branch 'main' into augusto/add-twmb-franz-go-integration

rarguelloF committed:
Merge remote-tracking branch 'origin/main' into augusto/add-twmb-franz-go-integration

genesor committed:
feat(franz-go): add orchestrion aspect for automatic instrumentation

genesor committed:
test(franz-go): add orchestrion integration test

Teko012 commented:
@rarguelloF @igoragoli Any progress on this by any chance? :)

igoragoli commented:
Hi @Teko012! We're waiting on an upstream contribution on twmb/franz-go which makes our integration much simpler to maintain: https://github.com/twmb/franz-go/pull/1289

rarguelloF committed:
refactor(franz-go): use hooks instead of wrapping (#4623)

<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Commit and PR titles should be prefixed with the general area of the
pull request's change.

-->
### What does this PR do?

<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succinct form,
consider
  opening multiple pull requests instead of a single one.
-->
Use the hooks added in https://github.com/twmb/franz-go/pull/1289 to
implement the integration without needing to return a custom type (see
https://github.com/DataDog/dd-trace-go/pull/4250/changes#r2952440108 for
more context).

Additional changes:
- Use instrumentation/testutils/containers for the integration tests.
- Remove unnecessary locking when reading the consumer group ID.
- Remove deprecated `WithAnalytics` options.
- Remove unnecessary build tags.

### Motivation

<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
* If this resolves a GitHub issue, include "Fixes #XXXX" to link the
issue and auto-close it on merge.
-->

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no
problems
during the review but the signing off is to be done by the reviewer(s).
-->

- [ ] Changed code has unit tests for its functionality at or near 100%
coverage.
- [ ] [System-Tests](https://github.com/DataDog/system-tests/) covering
this feature have been added and enabled with the va.b.c-dev version
tag.
- [ ] There is a benchmark for any new code, or changes to existing
code.
- [ ] If this interacts with the agent in a new way, a system test has
been added.
- [ ] New code is free of linting errors. You can check this by running
`make lint` locally.
- [ ] New code doesn't break existing tests. You can check this by
running `make test` locally.
- [ ] Add an appropriate team label so this PR gets put in the right
place for the release notes.
- [ ] All generated files are up to date. You can check this by running
`make generate` locally.
- [ ] Non-trivial go.mod changes, e.g. adding new modules, are reviewed
by @DataDog/dd-trace-go-guild. Make sure all nested modules are up to
date by running `make fix-modules` locally.

Unsure? Have a question? Request a review!

---------

Co-authored-by: Benjamin De Bernardi <benjamin.debernardi@datadoghq.com>

rarguelloF committed:
Merge branch 'main' into augusto/add-twmb-franz-go-integration

rarguelloF reviewed (approved):
LGTM, thanks for your patience!! 🙏

darccio commented:
@codex review

darccio commented on contrib/twmb/franz-go/carrier.go:
```diff
@@ -0,0 +1,53 @@
+	for i, h := range c.record.Headers {
+		if h.Key == key {
+			c.record.Headers[i].Value = []byte(val)
+			return
```
@igoragoli Does this make sense to you?

igoragoli replied:
> When a record already contains multiple headers for the same propagation key (franz-go stores headers as a slice, so this can happen when forwarding or reusing records)

Thank you, Dario! I have to check this assumption. I'll look into it later today :)

darccio reviewed (approved):
LGTM, although I'd address the Codex finding.

genesor committed:
chore(franz-go): register aspect in orchestrion/all

genesor committed:
fix(franz-go): remove unused createTopic from integration test

genesor committed:
test(franz-go): add ellipsis and no-args orchestrion aspect coverage

genesor committed:
fix(franz-go): use per-test topic/group to avoid shared-container pollution

genesor committed:
feat(franz-go): add orchestrion aspect for automatic instrumentation (#4629)

### What does this PR do?

Adds orchestrion support for the twmb/franz-go integration:

- Adds `orchestrion.yml` aspect that hooks into `kgo.NewClient()` and
appends `WithTracing()` to enable automatic distributed tracing
- Registers the aspect in `orchestrion/all`
- Adds an orchestrion integration test at
`internal/orchestrion/_integration/twmb_franz_go/`
- Adds the franz-go replace directive to the integration test module

### Motivation

Enable automatic instrumentation of franz-go Kafka clients via
orchestrion, matching the existing pattern used by other integrations
(gRPC, segmentio/kafka-go, etc.).

### Reviewer's Checklist

- [ ] Changed code has unit tests for its functionality at or near 100%
coverage.
- [ ] [System-Tests](https://github.com/DataDog/system-tests/) covering
this feature have been added and enabled with the va.b.c-dev version
tag.
- [ ] There is a benchmark for any new code, or changes to existing
code.
- [ ] If this interacts with the agent in a new way, a system test has
been added.
- [ ] New code is free of linting errors. You can check this by running
`make lint` locally.
- [ ] New code doesn't break existing tests. You can check this by
running `make test` locally.
- [ ] Add an appropriate team label so this PR gets put in the right
place for the release notes.
- [ ] All generated files are up to date. You can check this by running
`make generate` locally.
- [ ] Non-trivial go.mod changes, e.g. adding new modules, are reviewed
by @DataDog/dd-trace-go-guild. Make sure all nested modules are up to
date by running `make fix-modules` locally.

genesor reviewed (approved):
LGTM! Great work 🚀

igoragoli committed:
feat(franz-go): apply ForeachKey's handler only to the first occurence, apply Set to all occurences

igoragoli committed:
chore(franz-go): format and lint

Teko012 commented:
Do you know which version are you targeting this for?

genesor commented:
Hi @Teko012, we've included this PR in the ongoing `v2.8.0` release, https://github.com/DataDog/dd-trace-go/tree/release-v2.8.x

It's currently being tested internally and should be released soon.

jmsbtlr111 commented:
Looking forward to this addition! Thank you all who contributed.
